packages/gatsby-image/src/index.js

```diff
@@ -69,6 +76,95 @@ function getIO() {
   return io
 }
 
+function generateImageSources(imageVariants) {
+  const initialImage = imageVariants[0]
+
+  if (imageVariants.length === 1) {
+    return initialImage.srcSetWebp ? (
+      <source
+        type={`image/webp`}
+        media={initialImage.media}
+        srcSet={initialImage.srcSetWebp}
+        sizes={initialImage.sizes}
+      />
+    ) : null
+  }
+
+  return imageVariants.map(variant => {
+    const webPSource = variant.srcSetWebp ? (
+      <source
+        key={`${variant.src}-webp`}
+        type="image/webp"
+        media={variant.media}
+        srcSet={variant.srcSetWebp}
+        sizes={variant.sizes}
+      />
+    ) : (
+      undefined
+    )
+
+    const nonWebPSource = (
+      <source
+        key={variant.src}
+        media={variant.media}
+        srcSet={variant.srcSet}
+        sizes={variant.sizes}
+      />
+    )
+
+    return [webPSource, nonWebPSource]
+  })
+}
+
+function generateTracedSVGSources(imageVariants) {
+  if (imageVariants.length === 1) {
+    return null
+  }
+
+  return imageVariants.map(variant => (
+    <source
+      key={variant.src}
+      media={variant.media}
+      srcSet={variant.tracedSVG}
+    />
+  ))
+}
+
+function generateBase64Sources(imageVariants) {
+  if (imageVariants.length === 1) {
+    return null
+  }
+
+  return imageVariants.map(variant => (
+    <source key={variant.src} media={variant.media} srcSet={variant.base64} />
+  ))
+}
+
+function generateNoscriptSources(imageVariants, sizes) {
```

Comment:
I will refactor it to one of the ones you suggested! Good callout on `sizes`.

Regarding what happens if a `media` key is not passed: it will have the same behavior as the plain HTML `<picture>` tag would, which means the browser would just pick the first `<source>` it is capable of rendering. It does put the onus on the user to pass `media`, but no more so than `<picture>` itself does.